public/app/features/explore/spec/helper/setup.tsx

```diff
@@ -102,6 +102,24 @@ export function setupExplore(options?: SetupOptions): {
 
       return Promise.reject();
     },
+    getSync(datasource?: string | DataSourceRef | null): DataSourceApi | null {
+      let ds: DataSourceApi | undefined;
+      if (!datasource) {
+        ds = dsSettings[0]?.api;
+      } else {
+        ds = dsSettings.find((ds) =>
+          typeof datasource === 'string'
+            ? ds.api.name === datasource || ds.api.uid === datasource
+            : ds.api.uid === datasource?.uid
+        )?.api;
+      }
+
+      if (ds) {
+        return ds;
+      }
+
+      return null;
```

Comment:
I think this method should not be called `getSync` as it may lead to a wrong conclusion that it can be used interchangeably with `get`. Something like `getCachedDatasource` sounds more correct to me.

Also, I'm not sure if that's a good idea to make it part of the public API. Maybe it should be marked as internal? I think in Explore we can guarantee that for queries that were run and results passed to the visualization there's the corresponding data source loaded in the cache. With that assumption, visualization plugins can load the API synchronously, but I'm not sure if that assumption can be made in other places in Grafana. 

Also, the assumption that the data source is loaded in the cache when a plugin is run in Explore is an implementation detail that a plugin should not be aware of. I wonder if it'd make more sense to pass `getDatasource`  from Explore down to the visualization instead of calling the data source service directly from a plugin 🤔

Reply:
Agree with @ifrost  - I would rather turn this fn into a util function like `getPreloadedDataSource`. I can't recall any other places that would rely on requirements similar to the one you are describing.

Reply:
Sounds good. I'll close the PR until we've figured out a better solution for this. Thanks for your feedback, @dprokop and @ifrost  ! 👍